Lib/pdb.py

```diff
@@ -422,15 +424,28 @@ def interaction(self, frame, tb_or_exc):
         _exceptions = []
         if isinstance(tb_or_exc, BaseException):
             traceback, exception = tb_or_exc.__traceback__, tb_or_exc
-            current = exception
-            while current:
+            currents = [exception]
```

Comment:
I don't follow - what it ``currents``, and what is ``_exceptions``, and why do we need both?  Are you trying to allow navigating into exception groups? I think that would require some more design.

Reply:
Sorry I might have taken your following comment too literally:

> Which reminds me that we also need to handle cycles in the chain. Look at the traceback module to see how it's done there.

In the tb module you use a `queue` and a set of `_seen` traceback object ids. I assumed that what you wanted me to do, but the queue there can only old multiple elements because it handles exceptions group I belive. So I assumed you pointed me there for me to also see the exception group.

I've remove the handling of exception groups, and reverted to the previous logic, simply check if the cause or context is in the exceptions we've already visited.

Unless I still misunderstand what you asked.

Reply:
Sorry I wasn't clear. This is what I meant. This seems fine from the exceptions POV, but I'd like @gaogaotiantian to think about implications for PDB.   In particular, we are now saving a list of exceptions, which can create references to practically everything in the program through tracebacks (which reference frames). So what are the implications re memory leaks?

Reply:
One option is to store a traceback.TracebackException object instead of the exception itself. These are objects that don't hold references to the whole world.

Reply:
I don't think you can create weak refs to exceptions.

Reply:
During the PDB interaction, we are keeping all the frames anyway so I don't think having exceptions stored introduces any new issue. We should of course release the reference to the exceptions as long as we are out of PDB.

However, I do have some questions on this - why can't the release code be put in `forget()`? `setup()` calls `forget()` so we should extract the exception after `setup()`, other than that I did not see the calls to `forget()` during interaction, maybe I'm missing something?

Also, the code in `interaction` to get exceptions` is getting large here and I'd really like it to be another function.

It seems to me like `setup()` is doing very similar things. Should we have a new method like `setup_exceptions()` and call it in `setup()` if an exception is given? That's said, change the signature of `setup(self, f, tb)` to `setup(self, f, tb, exc=None)`, and do something like

```python
if exc is not None:
    self.setup_exceptions(exc)
```

in `setup()`.

I think this will make the `interaction()` function much clearer. Also we can do everything after the `forget()` in `setup()`. What do you think @iritkatriel ?

Reply:
> However, I do have some questions on this - why can't the release code be put in `forget()`? `setup()` calls `forget()` so we should extract the exception after `setup()`, other than that I did not see the calls to `forget()` during interaction, maybe I'm missing something?

Every time we `(pdb) exceptions <number>` to a new exception we call `setup(new_ex)`. If we put the cleanup code in `forget` then as soon as we jump to a new exception we'll loose all our setup. It is also called in `bp_commands` as well but I'm not sure the effect of that.

I'm happy to pull setup/interactions into two functions, but I'm slightly worried that this would have subtle backward compatibility with upstream. 

> Also we can do everything after the forget() in setup()

Again as is it will currently break, as setup is called when we jump exceptions, and if we split into say setup/setup_one_exception forget/forget_all_exception we still risk breaking if any subclass call setup/forget in the wrong place – I can tell that at least 2 Pdb derived debugger don't call self.setup()/self.forget()

Reply:
Adding onto this immediately, it looks like most setup/forget are in pairs, if we want to do a refactor that I would _prefer_ to do as a separate PR, I would change the setup/forget pair into a context manager.

I also pushed a commit that remove references to the exceptions after the command loop in interaction, and put a comment as to why it's not in forget.

Reply:
My bad, `setup()` is to setup the frames and `exceptions` needs to setup a new set of frames when switching. Never mind.

Can we still pack the code setting up the exceptions to a separate function and call it from `interaction`? The cleaning part is compact and easy to follow at this point, but the setup piece is a bit large.

I actually have a separate question for @iritkatriel - we expect this feature mainly used with `pm` right? What's the most likely usage from the users? I would guess it's executing the code with CLI - `python -m pdb the_code.py`. However, the current implementation won't work on it because the `main()` function is not changed for the new feature. Should we change the `main()` function?

Reply:
> Should we change the `main()` function?

I don't see why not.

Reply:
Then let's change the main function as well, so that `python -m pdb raise_exception.py` could enter pm with exception list. Another minor issue - when there's no exception, `exceptions` command does nothing. I think we should tell the user that there's no exceptions or something like that - so the users know that they are not supposed to use the command. I know `break` does the same thing - I don't like that either but let's deal with it later. We should do something like `display` - let the users know what happened.

Reply:
- Message added in do_exceptions if no chained exception found. 
- Main function modified to pass exception to post_mortem by default.
- Extracted the logic to create `self._chained_exceptions` into both a utility function and a context manager to make sure references are properly released.